### PR TITLE
[#808] improvement(spark): Verify the number of written records to enhance data correctness

### PR DESCRIPTION
**Pull request by rickyma**
### What changes were proposed in this pull request?

Verify the number of written records to enhance data accuracy.
Make sure all data records are sent by clients.
Make sure bugs like https://github.com/apache/incubator-uniffle/pull/714 will never be introduced into the code.

### Why are the changes needed?

A follow-up PR for https://github.com/apache/incubator-uniffle/pull/848.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

Existing UTs.

### CODE REVIEW

File: client-spark/spark2/src/main/java/org/apache/spark/shuffle/writer/RssShuffleWriter.java [hunk 1]
```diff
@@ -247,7 +247,9 @@ public void write(Iterator<Product2<K, V>> records) {
 
   private void writeImpl(Iterator<Product2<K, V>> records) {
     List<ShuffleBlockInfo> shuffleBlockInfos;
+    int recordCount = 0;
```
Comment (zuston) [inline client-spark/spark2/src/main/java/org/apache/spark/shuffle/writer/RssShuffleWriter.java:250]: `long` will be better that should be same with the `AtomicLong`

File: client-spark/spark3/src/main/java/org/apache/spark/shuffle/writer/RssShuffleWriter.java [hunk 2]
```diff
@@ -262,7 +262,9 @@ private void writeImpl(Iterator<Product2<K, V>> records) {
     if (isCombine) {
       createCombiner = shuffleDependency.aggregator().get().createCombiner();
     }
+    int recordCount = 0;
```
Comment (zuston) [inline client-spark/spark3/src/main/java/org/apache/spark/shuffle/writer/RssShuffleWriter.java:265]: ditto

File: client-spark/spark3/src/main/java/org/apache/spark/shuffle/writer/RssShuffleWriter.java [hunk 3]
```diff
@@ -285,6 +287,7 @@ private void writeImpl(Iterator<Product2<K, V>> records) {
       processShuffleBlockInfos(shuffleBlockInfos);
     }
     long checkStartTs = System.currentTimeMillis();
+    assert recordCount == bufferManager.getRecordCount();
```
Comment (zuston) [inline client-spark/spark3/src/main/java/org/apache/spark/shuffle/writer/RssShuffleWriter.java:290]: please throw new concret exception message here.

**Review by zuston — COMMENTED** [review]
LGTM overall. Left some minor comments.

And I hope this could be covered by test cases. Like throwing exception when encountering records is not same.

**Comment by rickyma** [comment]
> And I hope this could be covered by test cases. Like throwing exception when encountering records is not same.

I think the existing UTs have already covered this. Because this is a base method which will be used in many places.
If anything bad introduced into the code, I think the UTs won't pass.

File: client-spark/spark3/src/main/java/org/apache/spark/shuffle/writer/RssShuffleWriter.java (hunk 3 above)
Comment (rickyma) [reply client-spark/spark3/src/main/java/org/apache/spark/shuffle/writer/RssShuffleWriter.java]: Done.

File: client-spark/spark3/src/main/java/org/apache/spark/shuffle/writer/RssShuffleWriter.java (hunk 2 above)
Comment (rickyma) [reply client-spark/spark3/src/main/java/org/apache/spark/shuffle/writer/RssShuffleWriter.java]: Done.

File: client-spark/spark2/src/main/java/org/apache/spark/shuffle/writer/RssShuffleWriter.java (hunk 1 above)
Comment (rickyma) [reply client-spark/spark2/src/main/java/org/apache/spark/shuffle/writer/RssShuffleWriter.java]: Done.

**Comment by jerqi** [comment]
Maybe record count isn't enough. We should check block count, too.

**Comment by rickyma** [comment]
> Maybe record count isn't enough. We should check block count, too.

BlockIds will be checked in the method `checkBlockSendResult`. If any block is failed or lost, we will know.

**Comment by zuston** [comment]
> > And I hope this could be covered by test cases. Like throwing exception when encountering records is not same.
> 
> I think the existing UTs have already covered this. Because this is a base method which will be used in many places. If anything bad is introduced into the code in the future, I think the UTs won't pass.

No I hope you could mock throwing exception when record is not same.

**Review by zuston — APPROVED** [review]

**Comment by zuston** [comment]
Merged. Thanks @rickyma

**Pull request merged**